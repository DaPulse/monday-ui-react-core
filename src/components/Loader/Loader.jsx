import React, { forwardRef } from "react";
import PropTypes from "prop-types";
import "./Loader.scss";
import { backwardCompatibilityForProperties } from "../../helpers/backwardCompatibilityForProperties";

const Loader = forwardRef(
  (
    {
      // Backward compatibility for props naming
      svgClassName,
      className
    },
    ref
  ) => {
    const overrideClassName = backwardCompatibilityForProperties([className, svgClassName], "");
    return (
      <div className="monday-loader-component" ref={ref} role="alert" title="loading">
        <svg className={`circle-loader-spinner ${overrideClassName}`} viewBox="0 0 50 50">
          <circle className="circle-loader-spinner-path" cx="25" cy="25" r="20" fill="none" strokeWidth="5" />
        </svg>
      </div>
    );
  }
);

Loader.propTypes = {
  className: PropTypes.string
};

Loader.defaultProps = {
  className: undefined
};

export default Loader;
