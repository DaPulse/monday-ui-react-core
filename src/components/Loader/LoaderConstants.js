export const LOADER_COLORS = {
  PRIMARY: "var(--primary-color)",
  SECONDARY: "var(--secondary-text-color)",
  ON_PRIMARY: "var(--text-color-on-primary)",
  DARK: "var(--primary-text-color)"
};

export const LOADER_SIZES = {
  XS: "xs",
  SMALL: "small",
  MEDIUM: "medium",
  LARGE: "large"
};

export const LOADER_TYPES = {
  CASUAL: "casual",
  WITH_BACKGROUND: "with-background"
};
