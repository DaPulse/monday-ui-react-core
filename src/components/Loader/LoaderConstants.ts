export enum LoaderColors {
  PRIMARY = "var(--primary-color)",
  SECONDARY = "var(--secondary-text-color)",
  ON_PRIMARY = "var(--text-color-on-inverted)",
  DARK = "var(--primary-text-color)"
}

export const LoaderSizes = {
  XS: 16,
  SMALL: 24,
  MEDIUM: 40,
  LARGE: 64
};

export type LoaderSize = typeof LoaderSizes | number;
// export type LoaderSize = typeof LoaderSizes[keyof typeof LoaderSizes];

// export type LoaderSize = keyof typeof LOADER_SIZES;
