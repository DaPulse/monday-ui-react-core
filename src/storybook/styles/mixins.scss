@import "../../styles/themes";
@import "../../styles/global-css-settings";

@mixin frame() {
  border: var(--layout-border-color) 1px solid;
  border-radius: var(--border-radius-medium);
  @include theme-prop(border-color, layout-border-color);
}
