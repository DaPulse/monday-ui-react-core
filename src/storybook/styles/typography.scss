@import "../../styles/global-css-settings.scss";

//font family
$header-primary-font: "Sofia Pro", Roboto, sans-serif;
$text-primary-font: Roboto, sans-serif;

// sizes
$h1-font-size: 40px;
$h2-font-size: 28px;
$h3-font-size: 24px;
$anchor-font-size: 18px;
$text-font-size: 16px;
$small-text-font-size: 14px;

// line height
$text-line-height: 24px;

@mixin basic-text {
  @include theme-prop(color, primary-text-color);
}

@mixin content-text {
  @include basic-text;
  font-size: $text-font-size;
  line-height: $text-line-height;
}

@mixin paragraph-text {
  @include basic-text;
  font-size: $small-text-font-size;
  font-weight: 300;
  line-height: $text-line-height;
}