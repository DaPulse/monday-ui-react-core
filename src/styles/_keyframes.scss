// KEYFRAMES: This loader imports animation mixins content for reuse.

// Loading available keyframe content
@import "src/styles/keyframes/pop/pop-elastic";
@import "src/styles/keyframes/pop/pop-in";
@import "src/styles/keyframes/pop/pop-in-from-trigger";
@import "src/styles/keyframes/pop/pop-in-emphasized";
@import "src/styles/keyframes/pop/pop-in-elastic";
@import "src/styles/keyframes/pop/pop-in-elastic-bold";
@import "src/styles/keyframes/pop/pop-out-from-trigger";
@import "src/styles/keyframes/spin/spin-in-emphasized";
@import "src/styles/keyframes/slide/slide-in";
@import "src/styles/keyframes/slide/slide-in-elastic";
@import "src/styles/keyframes/slide/slide-out";

// MIXIN: Keyframe
// This mixin defines a keyframe definition using @content.
// @argument: $animation-name
// This mixin is used to enable reuse for local keyframes content

@mixin keyframe($animation_name) {
  @keyframes #{$animation_name} {
    @content;
  }
}