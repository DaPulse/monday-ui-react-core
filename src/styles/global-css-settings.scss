@import "./theme-mixin";

$box-shadow-small: 0 4px 8px -1px rgba(0, 0, 0, 0.2);
$box-shadow-medium: 0px 6px 20px -2px rgba(0, 0, 0, 0.2);
$box-shadow-large: 0px 15px 50px -10px rgba(0, 0, 0, 0.3);

$border-radius-extra-small: 2px;
$border-radius-small: 4px;
$border-radius-medium: 8px;
$border-radius-big: 16px;

$spacing-xxs: 2px;
$spacing-xs: 4px;
$spacing-xs-small: 6px;
$spacing-small: 8px;
$spacing-small-medium: 12px;
$spacing-medium: 16px;
$spacing-large: 24px;
$spacing-xl: 32px;
$spacing-xxl: 48px;
$spacing-xxxl: 64px;

$expand-animation-timing: cubic-bezier(0, 0, 0.35, 1);

@mixin box-shadow-xs() {
  @include theme-prop(box-shadow, box-shadow-xs);
}

@mixin box-shadow-small() {
  @include theme-prop(box-shadow, box-shadow-small);
}

@mixin box-shadow-medium() {
  @include theme-prop(box-shadow, box-shadow-medium);
}

@mixin box-shadow-large() {
  @include theme-prop(box-shadow, box-shadow-large);
}

@mixin hidden-element() {
  opacity: 0;
  width: 0;
  height: 0;
  position: absolute;
}

// Grid auto fit
// - This mixin defines a grid with auto fit repeat cells using min-max funtion.
// -- This allows grid to accommodate container width without media queries.
// -- it uses a repeat grid function with auto fix and minmax.
// -- grid adaptation is due to the min value along with auto-fit and 1fr max value.

// @params:
// - $grid-gap: null , defines grid "gap" attribute.
// - $grid-column-gap: null,  defines grid "column-gap" attribute.
// - $grid-row-gap: null,  defines grid "row-gap" attribute.
// - $grid-cell-min-width: {mandatory}, defined min value in for grid-template-columns

@mixin grid-auto-fit($grid-gap: null, $grid-column-gap: null, $grid-row-gap: null, $grid-cell-min-width) {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(#{$grid-cell-min-width}, 1fr));
  gap: $grid-gap;
  column-gap: $grid-column-gap;
  row-gap: $grid-row-gap;
}
