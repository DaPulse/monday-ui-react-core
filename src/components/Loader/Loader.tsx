import React, { ForwardedRef, forwardRef, useMemo } from "react";
import cx from "classnames";
import { backwardCompatibilityForProperties } from "../../helpers/backwardCompatibilityForProperties";
import { LoaderColors, LoaderSize, LoaderSizes } from "./LoaderConstants";
import styles from "./Loader.module.scss";

export interface ILoaderProps {
  id?: string;
  // Backward compatibility for props naming
  svgClassName?: string;
  className?: string;
  /** The loader's size: `number` or */
  size?: LoaderSize;
  color?: LoaderColors;
  hasBackground?: boolean;
}

const Loader: React.ForwardRefExoticComponent<ILoaderProps> & {
  sizes?: typeof LoaderSizes;
  colors?: typeof LoaderColors;
} = forwardRef<unknown, ILoaderProps>(
  ({ svgClassName, className, size, color, hasBackground, id }: ILoaderProps, ref: ForwardedRef<HTMLDivElement>) => {
    const overrideClassName = backwardCompatibilityForProperties([className, svgClassName], "");

    const sizeStyle = useMemo(() => {
      if (typeof size === "number") {
        return { width: size, height: size };
      }
      return undefined;
    }, [size]);

    return (
      <div
        className={cx("monday-loader-component", styles.loaderContainer)}
        ref={ref}
        role="alert"
        title="loading"
        style={sizeStyle}
        id={id}
      >
        <svg
          className={cx("circle-loader-spinner", styles.circleLoaderSpinner, overrideClassName)}
          viewBox="0 0 50 50"
          color={color}
          aria-hidden
        >
          {hasBackground && (
            <circle
              className={styles.circleLoaderSpinnerBackground}
              cx="25"
              cy="25"
              r="20"
              fill="none"
              strokeWidth="5"
            />
          )}
          <circle
            className={cx("circle-loader-spinner-path", styles.circleLoaderSpinnerPath)}
            cx="25"
            cy="25"
            r="20"
            fill="none"
            strokeWidth="5"
          />
        </svg>
      </div>
    );
  }
);

Object.assign(Loader, {
  sizes: LoaderSizes,
  colors: LoaderColors
});

Loader.defaultProps = {
  id: undefined,
  className: undefined,
  svgClassName: undefined,
  size: undefined,
  color: undefined,
  hasBackground: false
};

export default Loader;
