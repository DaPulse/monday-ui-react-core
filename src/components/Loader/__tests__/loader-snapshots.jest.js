import React from "react";
import renderer from "react-test-renderer";
import Loader from "../Loader";
import styles from "./loader-snapshots.jest.module.scss";

describe("Loader renders correctly", () => {
  it("with custom class name", () => {
    const tree = renderer.create(<Loader size={Loader.sizes.MEDIUM} className="dummy-class-name" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("with size variants", () => {
    it("with XS size", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.XS} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with SMALL size", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.SMALL} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with MEDIUM size", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.MEDIUM} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with LARGE size", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.LARGE} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with custom size from class", () => {
      const tree = renderer.create(<Loader className={styles.mondayLoaderComponent} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with custom size from parent", () => {
      const tree = renderer
        .create(
          <div style={{ width: 100, height: 100 }}>
            <Loader />
          </div>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("with color variants", () => {
    it("with PRIMARY color", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.MEDIUM} color={Loader.colors.PRIMARY} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with ON_PRIMARY color", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.MEDIUM} color={Loader.colors.ON_PRIMARY} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with SECONDARY color", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.MEDIUM} color={Loader.colors.SECONDARY} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with DARK color", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.MEDIUM} color={Loader.colors.DARK} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with custom color from parent", () => {
      const tree = renderer
        .create(
          <div style={{ color: "var(--color-dark-red)" }}>
            <Loader size={Loader.sizes.MEDIUM} />
          </div>
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe("with different types", () => {
    it("with CASUAL type", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.MEDIUM} type={Loader.types.CASUAL} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it("with WITH_BACKGROUND type", () => {
      const tree = renderer.create(<Loader size={Loader.sizes.MEDIUM} type={Loader.types.WITH_BACKGROUND} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
