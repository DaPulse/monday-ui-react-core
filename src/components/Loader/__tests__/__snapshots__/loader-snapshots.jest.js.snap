// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader renders correctly with color variants with DARK color 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    color="var(--primary-text-color)"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with color variants with ON_PRIMARY color 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    color="var(--text-color-on-inverted)"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with color variants with PRIMARY color 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    color="var(--primary-color)"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with color variants with SECONDARY color 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    color="var(--secondary-text-color)"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with color variants with custom color from parent 1`] = `
<div
  style={
    Object {
      "color": "var(--color-dark-red)",
    }
  }
>
  <div
    className="monday-loader-component"
    role="alert"
    style={
      Object {
        "height": 40,
        "width": 40,
      }
    }
    title="loading"
  >
    <svg
      className="circle-loader-spinner"
      viewBox="0 0 50 50"
    >
      <circle
        className="circle-loader-spinner-path"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        strokeWidth="5"
      />
    </svg>
  </div>
</div>
`;

exports[`Loader renders correctly with custom class name 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
  title="loading"
>
  <svg
    className="dummy-class-name circle-loader-spinner"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with different types with CASUAL type 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with different types with WITH_BACKGROUND type 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    viewBox="0 0 50 50"
  >
    <circle
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with size variants with LARGE size 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 64,
      "width": 64,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with size variants with MEDIUM size 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 40,
      "width": 40,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with size variants with SMALL size 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 24,
      "width": 24,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with size variants with XS size 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  style={
    Object {
      "height": 16,
      "width": 16,
    }
  }
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with size variants with custom size from class 1`] = `
<div
  className="monday-loader-component"
  role="alert"
  title="loading"
>
  <svg
    className="circle-loader-spinner"
    viewBox="0 0 50 50"
  >
    <circle
      className="circle-loader-spinner-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      strokeWidth="5"
    />
  </svg>
</div>
`;

exports[`Loader renders correctly with size variants with custom size from parent 1`] = `
<div
  style={
    Object {
      "height": 100,
      "width": 100,
    }
  }
>
  <div
    className="monday-loader-component"
    role="alert"
    title="loading"
  >
    <svg
      className="circle-loader-spinner"
      viewBox="0 0 50 50"
    >
      <circle
        className="circle-loader-spinner-path"
        cx="25"
        cy="25"
        fill="none"
        r="20"
        strokeWidth="5"
      />
    </svg>
  </div>
</div>
`;
